src/test/java/org/elasticsearch/index/query/ExistsQueryBuilderTest.java

```diff
@@ -0,0 +1,111 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.query;
+
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.ConstantScoreQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermRangeQuery;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.lucene.search.Queries;
+import org.elasticsearch.index.mapper.FieldMapper;
+import org.elasticsearch.index.mapper.internal.FieldNamesFieldMapper;
+
+import java.io.IOException;
+import java.util.Collection;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class ExistsQueryBuilderTest extends BaseQueryTestCase<ExistsQueryBuilder> {
+
+    @Override
+    protected ExistsQueryBuilder createEmptyQueryBuilder() {
+        return new ExistsQueryBuilder(null);
+    }
+
+    @Override
+    protected Query createExpectedQuery(ExistsQueryBuilder queryBuilder, QueryParseContext context) throws IOException {
+        final FieldNamesFieldMapper fieldNamesMapper = (FieldNamesFieldMapper)context.mapperService().fullName(FieldNamesFieldMapper.NAME);
+        String fieldNameInTest = queryBuilder.name();
+        if (getCurrentTypes().length > 0 && fieldNameInTest.equals(BaseQueryTestCase.OBJECT_FIELD_NAME)) {
+            // objects have two inner fields, add wildcard to match behaviour of extending field name pattern in test
+            fieldNameInTest = fieldNameInTest +".*";
+        }
+        Collection<String> fields = context.simpleMatchToIndexNames(fieldNameInTest);
+
+        if (fields.isEmpty()) {
+            return Queries.newMatchNoDocsQuery();
+        }
+
+        BooleanQuery boolFilter = new BooleanQuery();
+        for (String field : fields) {
+            if (fieldNamesMapper == null || !fieldNamesMapper.enabled()) {
+                FieldMapper mapper = context.fieldMapper(field);
+                if (mapper == null) {
+                    boolFilter.add(new TermRangeQuery(field, null, null, true, true), BooleanClause.Occur.SHOULD);
+                } else {
+                    boolFilter.add(mapper.rangeQuery(null, null, true, true, context), BooleanClause.Occur.SHOULD);
+                }
+            } else {
+                boolFilter.add(fieldNamesMapper.termQuery(field, context), BooleanClause.Occur.SHOULD);
+            }
+        }
+        return new ConstantScoreQuery(boolFilter);
+    }
+
+    @Override
+    protected void assertLuceneQuery(ExistsQueryBuilder queryBuilder, Query query, QueryParseContext context) {
+        if (queryBuilder.queryName() != null) {
+            Query namedQuery = context.copyNamedFilters().get(queryBuilder.queryName());
+            if (query instanceof ConstantScoreQuery) {
```

Comment:
I am always conflicted with this type of if based on what we are testing...shouldn't it be instead something around when we expect to have this type of query based on the query builder (the input of the test, not the output)?